blockchain/indexers/slpindex.go

```diff
@@ -0,0 +1,877 @@
+//   -----
+//   Max: 258 bytes
+// -----------------------------------------------------------------------------
+
+// TokenMetadata ...
+type TokenMetadata struct {
+	TokenID       *chainhash.Hash
+	SlpVersion    uint16
+	NftGroupID    *chainhash.Hash
+	MintBatonHash *chainhash.Hash
+	MintBatonVout uint32
+}
+
+// dbPutTokenIDIndexEntry uses an existing database transaction to update or add
+// the index entries for the hash to id and id to hash mappings for the provided
+// values.
+func dbPutTokenIDIndexEntry(dbTx database.Tx, id uint32, metadata *TokenMetadata) error {
+	// Serialize the height for use in the index entries.
+	var serializedID [4]byte
+	byteOrder.PutUint32(serializedID[:], id)
+
+	// Add the token ID by token hash mapping to the index.
+	meta := dbTx.Metadata()
+	hashIndex := meta.Bucket(tokenIDByHashIndexBucketName)
+	if err := hashIndex.Put(metadata.TokenID[:], serializedID[:]); err != nil {
+		return err
+	}
+
+	// Add or update token metadata by uint32 tokenID mapping to the index.
+	tmIndex := meta.Bucket(tokenMetadataByIDIndexBucketName)
+	tokenMetadata := make([]byte, 32+2+32+4)
+	copy(tokenMetadata[0:], metadata.TokenID[:])
+	byteOrder.PutUint16(tokenMetadata[32:], metadata.SlpVersion)
+	if metadata.NftGroupID != nil {
+		copy(tokenMetadata[34:], metadata.NftGroupID[:])
+		tokenMetadata = tokenMetadata[:66]
+	} else if metadata.MintBatonHash != nil {
+		copy(tokenMetadata[34:], metadata.MintBatonHash[:])
+		byteOrder.PutUint32(tokenMetadata[66:], metadata.MintBatonVout)
+	} else {
+		tokenMetadata = tokenMetadata[:34]
+	}
+
+	if metadata.NftGroupID == nil && metadata.SlpVersion == 0x41 {
+		panic("missing nft group id for NFT child " + string(id))
+	}
+
+	return tmIndex.Put(serializedID[:], tokenMetadata)
+}
+
+// dbFetchTokenIDByHash uses an existing database transaction to retrieve the
+// token id for the provided hash from the index.
+func dbFetchTokenIDByHash(dbTx database.Tx, hash *chainhash.Hash) (uint32, error) {
+	hashIndex := dbTx.Metadata().Bucket(tokenIDByHashIndexBucketName)
+	serializedID := hashIndex.Get(hash[:])
+	if serializedID == nil {
+		return 0, errNoTokenIDEntry
+	}
+	return byteOrder.Uint32(serializedID), nil
+}
+
+// dbFetchTokenMetadataBySerializedID uses an existing database transaction to
+// retrieve the hash for the provided serialized token id from the index.
+func dbFetchTokenMetadataBySerializedID(dbTx database.Tx, serializedID []byte) (*TokenMetadata, error) {
+	idIndex := dbTx.Metadata().Bucket(tokenMetadataByIDIndexBucketName)
+	serializedData := idIndex.Get(serializedID)
+	if serializedData == nil {
+		return nil, errNoTokenIDEntry
+	}
+
+	tokenIDHash, _ := chainhash.NewHash(serializedData[0:32])
+	slpVersion := byteOrder.Uint16(serializedData[32:34])
+
+	var (
+		mintBatonHash *chainhash.Hash
+		mintBatonVout uint32
+		nft1GroupID   *chainhash.Hash
+	)
+	if len(serializedData) == 70 {
+		if slpVersion == 0x41 {
+			panic("cannon have this length with nft1 child")
```

Comment:
NIT: spelling. Also, we should not panic here or we will cause a very nasty unexpected shutdown, we need to make sure it just bubbles up errors and shuts down if needed.

Reply:
done.

Reply:
Still a panic? Maybe a commit is missing? Seems both panics are not fixed, but I see a done comment.